Libraries/LibMedia/FFmpeg/FFmpegDemuxer.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright (c) 2025, Luke Wilde <luke@ladybird.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Error.h>
+#include <AK/NonnullOwnPtr.h>
+#include <LibMedia/Demuxer.h>
+#include <LibMedia/FFmpeg/FFmpegIOContext.h>
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+}
+
+namespace Media::FFmpeg {
+
+class FFmpegDemuxer : public Demuxer {
+public:
+    static ErrorOr<NonnullOwnPtr<FFmpegDemuxer>> create(NonnullOwnPtr<SeekableStream> stream);
+
+    FFmpegDemuxer(NonnullOwnPtr<SeekableStream> stream, NonnullOwnPtr<Media::FFmpeg::FFmpegIOContext>);
+    virtual ~FFmpegDemuxer() override;
+
+    virtual DecoderErrorOr<Vector<Track>> get_tracks_for_type(TrackType type) override;
+
+    virtual DecoderErrorOr<Optional<AK::Duration>> seek_to_most_recent_keyframe(Track track, AK::Duration timestamp, Optional<AK::Duration> earliest_available_sample = OptionalNone()) override;
+
+    virtual DecoderErrorOr<AK::Duration> duration(Track track) override;
+
+    virtual DecoderErrorOr<CodecID> get_codec_id_for_track(Track track) override;
+
+    virtual DecoderErrorOr<ReadonlyBytes> get_codec_initialization_data_for_track(Track track) override;
+
+    virtual DecoderErrorOr<Sample> get_next_sample_for_track(Track track) override;
+
+private:
+    ErrorOr<void> initialize();
```

Comment:
couple mixed initialization patterns going on with this class.

- Why do we need initialize() when we have the factory function create()? 
- Why does the constructor need to be public? Should anyone be creating this class directly without the factory? Won't calling the constructor externally mean that initialize is never called?

Reply:
- Moved initialisation into create
- The constructor is required to be public by `make<T>()`, unless there's a way around it I'm not seeing.